{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# caQTL 2\n",
    "\n",
    "We previously identified a proper subset of our data, let's now follow the rest of chrombpnet preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (10, 19)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>molecular_trait_id</th><th>chromosome</th><th>position</th><th>ref</th><th>alt</th><th>variant</th><th>ma_samples</th><th>maf</th><th>pvalue</th><th>beta</th><th>se</th><th>type</th><th>ac</th><th>an</th><th>r2</th><th>molecular_trait_object_id</th><th>gene_id</th><th>median_tpm</th><th>rsid</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;Interval_1&quot;</td><td>1</td><td>14464</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;chr1_14464_A_T…</td><td>48</td><td>0.263736</td><td>0.0712853</td><td>-0.328458</td><td>0.179592</td><td>&quot;SNP&quot;</td><td>48</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_1&quot;</td><td>&quot;Interval_1&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs546169444&quot;</td></tr><tr><td>&quot;Interval_2&quot;</td><td>1</td><td>14464</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;chr1_14464_A_T…</td><td>48</td><td>0.263736</td><td>0.167217</td><td>0.287697</td><td>0.206328</td><td>&quot;SNP&quot;</td><td>48</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_2&quot;</td><td>&quot;Interval_2&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs546169444&quot;</td></tr><tr><td>&quot;Interval_3&quot;</td><td>1</td><td>14464</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;chr1_14464_A_T…</td><td>48</td><td>0.263736</td><td>0.480256</td><td>0.108094</td><td>0.152387</td><td>&quot;SNP&quot;</td><td>48</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_3&quot;</td><td>&quot;Interval_3&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs546169444&quot;</td></tr><tr><td>&quot;Interval_4&quot;</td><td>1</td><td>14464</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;chr1_14464_A_T…</td><td>48</td><td>0.263736</td><td>0.805997</td><td>-0.042729</td><td>0.173386</td><td>&quot;SNP&quot;</td><td>48</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_4&quot;</td><td>&quot;Interval_4&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs546169444&quot;</td></tr><tr><td>&quot;Interval_5&quot;</td><td>1</td><td>14464</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;chr1_14464_A_T…</td><td>48</td><td>0.263736</td><td>0.206241</td><td>0.222002</td><td>0.174157</td><td>&quot;SNP&quot;</td><td>48</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_5&quot;</td><td>&quot;Interval_5&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs546169444&quot;</td></tr><tr><td>&quot;Interval_6&quot;</td><td>1</td><td>14464</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;chr1_14464_A_T…</td><td>48</td><td>0.263736</td><td>0.970258</td><td>0.005194</td><td>0.13886</td><td>&quot;SNP&quot;</td><td>48</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_6&quot;</td><td>&quot;Interval_6&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs546169444&quot;</td></tr><tr><td>&quot;Interval_1&quot;</td><td>1</td><td>14933</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>&quot;chr1_14933_G_A…</td><td>13</td><td>0.0714286</td><td>0.538671</td><td>0.163632</td><td>0.264955</td><td>&quot;SNP&quot;</td><td>13</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_1&quot;</td><td>&quot;Interval_1&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs199856693&quot;</td></tr><tr><td>&quot;Interval_2&quot;</td><td>1</td><td>14933</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>&quot;chr1_14933_G_A…</td><td>13</td><td>0.0714286</td><td>0.251985</td><td>-0.346142</td><td>0.299896</td><td>&quot;SNP&quot;</td><td>13</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_2&quot;</td><td>&quot;Interval_2&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs199856693&quot;</td></tr><tr><td>&quot;Interval_3&quot;</td><td>1</td><td>14933</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>&quot;chr1_14933_G_A…</td><td>13</td><td>0.0714286</td><td>0.07458</td><td>0.391881</td><td>0.2168</td><td>&quot;SNP&quot;</td><td>13</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_3&quot;</td><td>&quot;Interval_3&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs199856693&quot;</td></tr><tr><td>&quot;Interval_4&quot;</td><td>1</td><td>14933</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>&quot;chr1_14933_G_A…</td><td>13</td><td>0.0714286</td><td>0.0234297</td><td>0.561551</td><td>0.242845</td><td>&quot;SNP&quot;</td><td>13</td><td>182</td><td>&quot;NA&quot;</td><td>&quot;Interval_4&quot;</td><td>&quot;Interval_4&quot;</td><td>&quot;NA&quot;</td><td>&quot;rs199856693&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (10, 19)\n",
       "┌────────────┬────────────┬──────────┬─────┬───┬────────────┬────────────┬────────────┬────────────┐\n",
       "│ molecular_ ┆ chromosome ┆ position ┆ ref ┆ … ┆ molecular_ ┆ gene_id    ┆ median_tpm ┆ rsid       │\n",
       "│ trait_id   ┆ ---        ┆ ---      ┆ --- ┆   ┆ trait_obje ┆ ---        ┆ ---        ┆ ---        │\n",
       "│ ---        ┆ i64        ┆ i64      ┆ str ┆   ┆ ct_id      ┆ str        ┆ str        ┆ str        │\n",
       "│ str        ┆            ┆          ┆     ┆   ┆ ---        ┆            ┆            ┆            │\n",
       "│            ┆            ┆          ┆     ┆   ┆ str        ┆            ┆            ┆            │\n",
       "╞════════════╪════════════╪══════════╪═════╪═══╪════════════╪════════════╪════════════╪════════════╡\n",
       "│ Interval_1 ┆ 1          ┆ 14464    ┆ A   ┆ … ┆ Interval_1 ┆ Interval_1 ┆ NA         ┆ rs54616944 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 4          │\n",
       "│ Interval_2 ┆ 1          ┆ 14464    ┆ A   ┆ … ┆ Interval_2 ┆ Interval_2 ┆ NA         ┆ rs54616944 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 4          │\n",
       "│ Interval_3 ┆ 1          ┆ 14464    ┆ A   ┆ … ┆ Interval_3 ┆ Interval_3 ┆ NA         ┆ rs54616944 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 4          │\n",
       "│ Interval_4 ┆ 1          ┆ 14464    ┆ A   ┆ … ┆ Interval_4 ┆ Interval_4 ┆ NA         ┆ rs54616944 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 4          │\n",
       "│ …          ┆ …          ┆ …        ┆ …   ┆ … ┆ …          ┆ …          ┆ …          ┆ …          │\n",
       "│ Interval_1 ┆ 1          ┆ 14933    ┆ G   ┆ … ┆ Interval_1 ┆ Interval_1 ┆ NA         ┆ rs19985669 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 3          │\n",
       "│ Interval_2 ┆ 1          ┆ 14933    ┆ G   ┆ … ┆ Interval_2 ┆ Interval_2 ┆ NA         ┆ rs19985669 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 3          │\n",
       "│ Interval_3 ┆ 1          ┆ 14933    ┆ G   ┆ … ┆ Interval_3 ┆ Interval_3 ┆ NA         ┆ rs19985669 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 3          │\n",
       "│ Interval_4 ┆ 1          ┆ 14933    ┆ G   ┆ … ┆ Interval_4 ┆ Interval_4 ┆ NA         ┆ rs19985669 │\n",
       "│            ┆            ┆          ┆     ┆   ┆            ┆            ┆            ┆ 3          │\n",
       "└────────────┴────────────┴──────────┴─────┴───┴────────────┴────────────┴────────────┴────────────┘"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#chrombpnet subsets to only peaks, we can have one that is generated as peaks vs not I guess?\n",
    "#do they also only subset to the peak that the QTL is in? let's see\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "data_sus = pl.scan_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset.tsv', separator='\\t', infer_schema_length=1000)\n",
    "data_sus.fetch(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#so the intervals are these set peaks I think\n",
    "#and we can get the overlapping peaks with the metadata file which has the location of the peaks!!\n",
    "data_sus = (\n",
    "    data_sus\n",
    "    .with_columns(     # create a new column lazily\n",
    "        (pl.col(\"position\") - 1).alias(\"position0\")\n",
    "    )\n",
    "    .select([          # select only these columns\n",
    "        \"chromosome\", \"position0\", \"position\", \"ref\", \"alt\", \"variant\",\n",
    "        \"ma_samples\", \"maf\", \"pvalue\", \"beta\", \"se\", \"type\", \"ac\", \"an\", \"r2\",\n",
    "        \"molecular_trait_object_id\", \"gene_id\", \"median_tpm\", \"rsid\",\n",
    "        \"molecular_trait_id\"\n",
    "    ])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_72900/2038584751.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  peaks_data = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018_peak_metadata.tsv', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>quant_id</th>\n",
       "      <th>group_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>phenotype_pos</th>\n",
       "      <th>phenotype_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>10326</td>\n",
       "      <td>584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>29180</td>\n",
       "      <td>512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>181138</td>\n",
       "      <td>780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>183591</td>\n",
       "      <td>1893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>186738</td>\n",
       "      <td>744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249196</th>\n",
       "      <td>Interval_249197</td>\n",
       "      <td>Interval_249197</td>\n",
       "      <td>Interval_249197</td>\n",
       "      <td>Interval_249197</td>\n",
       "      <td>Y</td>\n",
       "      <td>56861329</td>\n",
       "      <td>56862459</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_249197</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>56861894</td>\n",
       "      <td>1131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249197</th>\n",
       "      <td>Interval_249198</td>\n",
       "      <td>Interval_249198</td>\n",
       "      <td>Interval_249198</td>\n",
       "      <td>Interval_249198</td>\n",
       "      <td>Y</td>\n",
       "      <td>56863235</td>\n",
       "      <td>56864760</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_249198</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>56863998</td>\n",
       "      <td>1526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249198</th>\n",
       "      <td>Interval_249199</td>\n",
       "      <td>Interval_249199</td>\n",
       "      <td>Interval_249199</td>\n",
       "      <td>Interval_249199</td>\n",
       "      <td>Y</td>\n",
       "      <td>56865516</td>\n",
       "      <td>56874990</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_249199</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>56870253</td>\n",
       "      <td>9475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249199</th>\n",
       "      <td>Interval_249200</td>\n",
       "      <td>Interval_249200</td>\n",
       "      <td>Interval_249200</td>\n",
       "      <td>Interval_249200</td>\n",
       "      <td>Y</td>\n",
       "      <td>56876239</td>\n",
       "      <td>56880531</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_249200</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>56878385</td>\n",
       "      <td>4293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249200</th>\n",
       "      <td>Interval_249201</td>\n",
       "      <td>Interval_249201</td>\n",
       "      <td>Interval_249201</td>\n",
       "      <td>Interval_249201</td>\n",
       "      <td>Y</td>\n",
       "      <td>56884100</td>\n",
       "      <td>56885316</td>\n",
       "      <td>1</td>\n",
       "      <td>Interval_249201</td>\n",
       "      <td>open_chromatin</td>\n",
       "      <td>56884708</td>\n",
       "      <td>1217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249201 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           phenotype_id         quant_id         group_id          gene_id  \\\n",
       "0            Interval_1       Interval_1       Interval_1       Interval_1   \n",
       "1            Interval_2       Interval_2       Interval_2       Interval_2   \n",
       "2            Interval_3       Interval_3       Interval_3       Interval_3   \n",
       "3            Interval_4       Interval_4       Interval_4       Interval_4   \n",
       "4            Interval_5       Interval_5       Interval_5       Interval_5   \n",
       "...                 ...              ...              ...              ...   \n",
       "249196  Interval_249197  Interval_249197  Interval_249197  Interval_249197   \n",
       "249197  Interval_249198  Interval_249198  Interval_249198  Interval_249198   \n",
       "249198  Interval_249199  Interval_249199  Interval_249199  Interval_249199   \n",
       "249199  Interval_249200  Interval_249200  Interval_249200  Interval_249200   \n",
       "249200  Interval_249201  Interval_249201  Interval_249201  Interval_249201   \n",
       "\n",
       "       chromosome  gene_start  gene_end  strand        gene_name  \\\n",
       "0               1       10034     10617       1       Interval_1   \n",
       "1               1       28924     29435       1       Interval_2   \n",
       "2               1      180749    181528       1       Interval_3   \n",
       "3               1      182645    184537       1       Interval_4   \n",
       "4               1      186367    187110       1       Interval_5   \n",
       "...           ...         ...       ...     ...              ...   \n",
       "249196          Y    56861329  56862459       1  Interval_249197   \n",
       "249197          Y    56863235  56864760       1  Interval_249198   \n",
       "249198          Y    56865516  56874990       1  Interval_249199   \n",
       "249199          Y    56876239  56880531       1  Interval_249200   \n",
       "249200          Y    56884100  56885316       1  Interval_249201   \n",
       "\n",
       "             gene_type  phenotype_pos  phenotype_length  \n",
       "0       open_chromatin          10326               584  \n",
       "1       open_chromatin          29180               512  \n",
       "2       open_chromatin         181138               780  \n",
       "3       open_chromatin         183591              1893  \n",
       "4       open_chromatin         186738               744  \n",
       "...                ...            ...               ...  \n",
       "249196  open_chromatin       56861894              1131  \n",
       "249197  open_chromatin       56863998              1526  \n",
       "249198  open_chromatin       56870253              9475  \n",
       "249199  open_chromatin       56878385              4293  \n",
       "249200  open_chromatin       56884708              1217  \n",
       "\n",
       "[249201 rows x 12 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_data = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018_peak_metadata.tsv', sep='\\t')\n",
    "peaks_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_72900/3427181243.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  peaks_data[\"chromosome\"] = \"chr\"+peaks_data[\"chromosome\"] .astype(str)\n",
      "/tmp/ipykernel_72900/3427181243.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  peaks_data[\"center\"]  =  ((peaks_data[\"gene_start\"] + peaks_data[\"gene_end\"])/2).astype(int)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249196</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56861329</td>\n",
       "      <td>56862459</td>\n",
       "      <td>Interval_249197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56863235</td>\n",
       "      <td>56864760</td>\n",
       "      <td>Interval_249198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249198</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56865516</td>\n",
       "      <td>56874990</td>\n",
       "      <td>Interval_249199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249199</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56876239</td>\n",
       "      <td>56880531</td>\n",
       "      <td>Interval_249200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249200</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56884100</td>\n",
       "      <td>56885316</td>\n",
       "      <td>Interval_249201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249201 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome  gene_start  gene_end     phenotype_id\n",
       "0            chr1       10034     10617       Interval_1\n",
       "1            chr1       28924     29435       Interval_2\n",
       "2            chr1      180749    181528       Interval_3\n",
       "3            chr1      182645    184537       Interval_4\n",
       "4            chr1      186367    187110       Interval_5\n",
       "...           ...         ...       ...              ...\n",
       "249196       chrY    56861329  56862459  Interval_249197\n",
       "249197       chrY    56863235  56864760  Interval_249198\n",
       "249198       chrY    56865516  56874990  Interval_249199\n",
       "249199       chrY    56876239  56880531  Interval_249200\n",
       "249200       chrY    56884100  56885316  Interval_249201\n",
       "\n",
       "[249201 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_data = peaks_data[[\"chromosome\", \"gene_start\", \"gene_end\", \"phenotype_pos\", \"phenotype_id\"]]\n",
    "peaks_data[\"chromosome\"] = \"chr\"+peaks_data[\"chromosome\"] .astype(str)\n",
    "# peaks_data[\"center\"]  =  ((peaks_data[\"gene_start\"] + peaks_data[\"gene_end\"])/2).astype(int)\n",
    "# peaks_data[\"start\"] = peaks_data[\"center\"] - 100\n",
    "# peaks_data[\"end\"] = peaks_data[\"center\"] + 100\n",
    "peaks_data = peaks_data[['chromosome', 'gene_start', 'gene_end', \"phenotype_id\"]]\n",
    "\n",
    "#literally just make it so that we subset it. Then we get chromosome and call it chrX for whatever chromosome. Then we get center by doing start and end and averaging\n",
    "#then we define start and end as 100 away form the peaks... not perfect but I guess it's fine?\n",
    "peaks_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>type</th>\n",
       "      <th>ac</th>\n",
       "      <th>an</th>\n",
       "      <th>r2</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Interval_1</td>\n",
       "      <td>1</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.071285</td>\n",
       "      <td>-0.328458</td>\n",
       "      <td>0.179592</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Interval_2</td>\n",
       "      <td>1</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.287697</td>\n",
       "      <td>0.206328</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Interval_3</td>\n",
       "      <td>1</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.480256</td>\n",
       "      <td>0.108094</td>\n",
       "      <td>0.152387</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Interval_4</td>\n",
       "      <td>1</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.805997</td>\n",
       "      <td>-0.042729</td>\n",
       "      <td>0.173386</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Interval_5</td>\n",
       "      <td>1</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.206241</td>\n",
       "      <td>0.222002</td>\n",
       "      <td>0.174157</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Interval_4</td>\n",
       "      <td>1</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.499022</td>\n",
       "      <td>-0.154622</td>\n",
       "      <td>0.227640</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Interval_5</td>\n",
       "      <td>1</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.275051</td>\n",
       "      <td>-0.252686</td>\n",
       "      <td>0.229856</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Interval_6</td>\n",
       "      <td>1</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.421628</td>\n",
       "      <td>-0.147053</td>\n",
       "      <td>0.182015</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Interval_7</td>\n",
       "      <td>1</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.447159</td>\n",
       "      <td>-0.203737</td>\n",
       "      <td>0.266646</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_7</td>\n",
       "      <td>Interval_7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Interval_8</td>\n",
       "      <td>1</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.555134</td>\n",
       "      <td>0.142618</td>\n",
       "      <td>0.240634</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_8</td>\n",
       "      <td>Interval_8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   molecular_trait_id  chromosome  position ref alt          variant  \\\n",
       "0          Interval_1           1     14464   A   T   chr1_14464_A_T   \n",
       "1          Interval_2           1     14464   A   T   chr1_14464_A_T   \n",
       "2          Interval_3           1     14464   A   T   chr1_14464_A_T   \n",
       "3          Interval_4           1     14464   A   T   chr1_14464_A_T   \n",
       "4          Interval_5           1     14464   A   T   chr1_14464_A_T   \n",
       "..                ...         ...       ...  ..  ..              ...   \n",
       "95         Interval_4           1    267754   T   C  chr1_267754_T_C   \n",
       "96         Interval_5           1    267754   T   C  chr1_267754_T_C   \n",
       "97         Interval_6           1    267754   T   C  chr1_267754_T_C   \n",
       "98         Interval_7           1    267754   T   C  chr1_267754_T_C   \n",
       "99         Interval_8           1    267754   T   C  chr1_267754_T_C   \n",
       "\n",
       "    ma_samples       maf    pvalue      beta        se type  ac   an  r2  \\\n",
       "0           48  0.263736  0.071285 -0.328458  0.179592  SNP  48  182 NaN   \n",
       "1           48  0.263736  0.167217  0.287697  0.206328  SNP  48  182 NaN   \n",
       "2           48  0.263736  0.480256  0.108094  0.152387  SNP  48  182 NaN   \n",
       "3           48  0.263736  0.805997 -0.042729  0.173386  SNP  48  182 NaN   \n",
       "4           48  0.263736  0.206241  0.222002  0.174157  SNP  48  182 NaN   \n",
       "..         ...       ...       ...       ...       ...  ...  ..  ...  ..   \n",
       "95          16  0.087912  0.499022 -0.154622  0.227640  SNP  16  182 NaN   \n",
       "96          16  0.087912  0.275051 -0.252686  0.229856  SNP  16  182 NaN   \n",
       "97          16  0.087912  0.421628 -0.147053  0.182015  SNP  16  182 NaN   \n",
       "98          16  0.087912  0.447159 -0.203737  0.266646  SNP  16  182 NaN   \n",
       "99          16  0.087912  0.555134  0.142618  0.240634  SNP  16  182 NaN   \n",
       "\n",
       "   molecular_trait_object_id     gene_id  median_tpm         rsid  \n",
       "0                 Interval_1  Interval_1         NaN  rs546169444  \n",
       "1                 Interval_2  Interval_2         NaN  rs546169444  \n",
       "2                 Interval_3  Interval_3         NaN  rs546169444  \n",
       "3                 Interval_4  Interval_4         NaN  rs546169444  \n",
       "4                 Interval_5  Interval_5         NaN  rs546169444  \n",
       "..                       ...         ...         ...          ...  \n",
       "95                Interval_4  Interval_4         NaN  rs367639380  \n",
       "96                Interval_5  Interval_5         NaN  rs367639380  \n",
       "97                Interval_6  Interval_6         NaN  rs367639380  \n",
       "98                Interval_7  Interval_7         NaN  rs367639380  \n",
       "99                Interval_8  Interval_8         NaN  rs367639380  \n",
       "\n",
       "[100 rows x 19 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the next step is to merge the dataframes. We will do it by getting a subset of the data\n",
    "data_sus = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset.tsv', sep='\\t', nrows=100)\n",
    "data_sus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position0</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>type</th>\n",
       "      <th>ac</th>\n",
       "      <th>an</th>\n",
       "      <th>r2</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.071285</td>\n",
       "      <td>-0.328458</td>\n",
       "      <td>0.179592</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.287697</td>\n",
       "      <td>0.206328</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.480256</td>\n",
       "      <td>0.108094</td>\n",
       "      <td>0.152387</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.805997</td>\n",
       "      <td>-0.042729</td>\n",
       "      <td>0.173386</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.206241</td>\n",
       "      <td>0.222002</td>\n",
       "      <td>0.174157</td>\n",
       "      <td>SNP</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>1</td>\n",
       "      <td>267753</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.499022</td>\n",
       "      <td>-0.154622</td>\n",
       "      <td>0.227640</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>1</td>\n",
       "      <td>267753</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.275051</td>\n",
       "      <td>-0.252686</td>\n",
       "      <td>0.229856</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>1</td>\n",
       "      <td>267753</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.421628</td>\n",
       "      <td>-0.147053</td>\n",
       "      <td>0.182015</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "      <td>Interval_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>1</td>\n",
       "      <td>267753</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.447159</td>\n",
       "      <td>-0.203737</td>\n",
       "      <td>0.266646</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_7</td>\n",
       "      <td>Interval_7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "      <td>Interval_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1</td>\n",
       "      <td>267753</td>\n",
       "      <td>267754</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_267754_T_C</td>\n",
       "      <td>16</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>0.555134</td>\n",
       "      <td>0.142618</td>\n",
       "      <td>0.240634</td>\n",
       "      <td>SNP</td>\n",
       "      <td>16</td>\n",
       "      <td>182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_8</td>\n",
       "      <td>Interval_8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs367639380</td>\n",
       "      <td>Interval_8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome  position0  position ref alt          variant  ma_samples  \\\n",
       "0            1      14463     14464   A   T   chr1_14464_A_T          48   \n",
       "1            1      14463     14464   A   T   chr1_14464_A_T          48   \n",
       "2            1      14463     14464   A   T   chr1_14464_A_T          48   \n",
       "3            1      14463     14464   A   T   chr1_14464_A_T          48   \n",
       "4            1      14463     14464   A   T   chr1_14464_A_T          48   \n",
       "..         ...        ...       ...  ..  ..              ...         ...   \n",
       "95           1     267753    267754   T   C  chr1_267754_T_C          16   \n",
       "96           1     267753    267754   T   C  chr1_267754_T_C          16   \n",
       "97           1     267753    267754   T   C  chr1_267754_T_C          16   \n",
       "98           1     267753    267754   T   C  chr1_267754_T_C          16   \n",
       "99           1     267753    267754   T   C  chr1_267754_T_C          16   \n",
       "\n",
       "         maf    pvalue      beta        se type  ac   an  r2  \\\n",
       "0   0.263736  0.071285 -0.328458  0.179592  SNP  48  182 NaN   \n",
       "1   0.263736  0.167217  0.287697  0.206328  SNP  48  182 NaN   \n",
       "2   0.263736  0.480256  0.108094  0.152387  SNP  48  182 NaN   \n",
       "3   0.263736  0.805997 -0.042729  0.173386  SNP  48  182 NaN   \n",
       "4   0.263736  0.206241  0.222002  0.174157  SNP  48  182 NaN   \n",
       "..       ...       ...       ...       ...  ...  ..  ...  ..   \n",
       "95  0.087912  0.499022 -0.154622  0.227640  SNP  16  182 NaN   \n",
       "96  0.087912  0.275051 -0.252686  0.229856  SNP  16  182 NaN   \n",
       "97  0.087912  0.421628 -0.147053  0.182015  SNP  16  182 NaN   \n",
       "98  0.087912  0.447159 -0.203737  0.266646  SNP  16  182 NaN   \n",
       "99  0.087912  0.555134  0.142618  0.240634  SNP  16  182 NaN   \n",
       "\n",
       "   molecular_trait_object_id     gene_id  median_tpm         rsid  \\\n",
       "0                 Interval_1  Interval_1         NaN  rs546169444   \n",
       "1                 Interval_2  Interval_2         NaN  rs546169444   \n",
       "2                 Interval_3  Interval_3         NaN  rs546169444   \n",
       "3                 Interval_4  Interval_4         NaN  rs546169444   \n",
       "4                 Interval_5  Interval_5         NaN  rs546169444   \n",
       "..                       ...         ...         ...          ...   \n",
       "95                Interval_4  Interval_4         NaN  rs367639380   \n",
       "96                Interval_5  Interval_5         NaN  rs367639380   \n",
       "97                Interval_6  Interval_6         NaN  rs367639380   \n",
       "98                Interval_7  Interval_7         NaN  rs367639380   \n",
       "99                Interval_8  Interval_8         NaN  rs367639380   \n",
       "\n",
       "   molecular_trait_id  \n",
       "0          Interval_1  \n",
       "1          Interval_2  \n",
       "2          Interval_3  \n",
       "3          Interval_4  \n",
       "4          Interval_5  \n",
       "..                ...  \n",
       "95         Interval_4  \n",
       "96         Interval_5  \n",
       "97         Interval_6  \n",
       "98         Interval_7  \n",
       "99         Interval_8  \n",
       "\n",
       "[100 rows x 20 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_sus[\"position0\"] = data_sus[\"position\"] - 1\n",
    "data_sus = data_sus[['chromosome', 'position0', 'position', 'ref', 'alt', 'variant',\n",
    "       'ma_samples', 'maf', 'pvalue', 'beta', 'se', 'type', 'ac', 'an', 'r2',\n",
    "       'molecular_trait_object_id', 'gene_id', 'median_tpm', 'rsid',\n",
    "       'molecular_trait_id']]\n",
    "data_sus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_x</th>\n",
       "      <th>position0</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>...</th>\n",
       "      <th>r2</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.071285</td>\n",
       "      <td>-0.328458</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.287697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.480256</td>\n",
       "      <td>0.108094</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.805997</td>\n",
       "      <td>-0.042729</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.206241</td>\n",
       "      <td>0.222002</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.970258</td>\n",
       "      <td>0.005194</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>190802</td>\n",
       "      <td>191922</td>\n",
       "      <td>Interval_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.538671</td>\n",
       "      <td>0.163632</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.251985</td>\n",
       "      <td>-0.346142</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.074580</td>\n",
       "      <td>0.391881</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.023430</td>\n",
       "      <td>0.561551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome_x  position0  position ref alt         variant  ma_samples  \\\n",
       "0             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "1             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "2             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "3             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "4             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "5             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "6             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "7             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "8             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "9             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "\n",
       "        maf    pvalue      beta  ...  r2 molecular_trait_object_id  \\\n",
       "0  0.263736  0.071285 -0.328458  ... NaN                Interval_1   \n",
       "1  0.263736  0.167217  0.287697  ... NaN                Interval_2   \n",
       "2  0.263736  0.480256  0.108094  ... NaN                Interval_3   \n",
       "3  0.263736  0.805997 -0.042729  ... NaN                Interval_4   \n",
       "4  0.263736  0.206241  0.222002  ... NaN                Interval_5   \n",
       "5  0.263736  0.970258  0.005194  ... NaN                Interval_6   \n",
       "6  0.071429  0.538671  0.163632  ... NaN                Interval_1   \n",
       "7  0.071429  0.251985 -0.346142  ... NaN                Interval_2   \n",
       "8  0.071429  0.074580  0.391881  ... NaN                Interval_3   \n",
       "9  0.071429  0.023430  0.561551  ... NaN                Interval_4   \n",
       "\n",
       "      gene_id  median_tpm         rsid molecular_trait_id chromosome_y  \\\n",
       "0  Interval_1         NaN  rs546169444         Interval_1         chr1   \n",
       "1  Interval_2         NaN  rs546169444         Interval_2         chr1   \n",
       "2  Interval_3         NaN  rs546169444         Interval_3         chr1   \n",
       "3  Interval_4         NaN  rs546169444         Interval_4         chr1   \n",
       "4  Interval_5         NaN  rs546169444         Interval_5         chr1   \n",
       "5  Interval_6         NaN  rs546169444         Interval_6         chr1   \n",
       "6  Interval_1         NaN  rs199856693         Interval_1         chr1   \n",
       "7  Interval_2         NaN  rs199856693         Interval_2         chr1   \n",
       "8  Interval_3         NaN  rs199856693         Interval_3         chr1   \n",
       "9  Interval_4         NaN  rs199856693         Interval_4         chr1   \n",
       "\n",
       "   gene_start gene_end phenotype_id  \n",
       "0       10034    10617   Interval_1  \n",
       "1       28924    29435   Interval_2  \n",
       "2      180749   181528   Interval_3  \n",
       "3      182645   184537   Interval_4  \n",
       "4      186367   187110   Interval_5  \n",
       "5      190802   191922   Interval_6  \n",
       "6       10034    10617   Interval_1  \n",
       "7       28924    29435   Interval_2  \n",
       "8      180749   181528   Interval_3  \n",
       "9      182645   184537   Interval_4  \n",
       "\n",
       "[10 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_dataframe = data_sus.merge(peaks_data, right_on=['phenotype_id'], left_on=[\"molecular_trait_id\"], how='left')\n",
    "merged_dataframe.head(10)\n",
    "#so merges with the interval, so interval 1 when it repeats we see that same information, basically gives the chromosome and gene start and end for each one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_x</th>\n",
       "      <th>position0</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>...</th>\n",
       "      <th>r2</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chromosome_x, position0, position, ref, alt, variant, ma_samples, maf, pvalue, beta, se, type, ac, an, r2, molecular_trait_object_id, gene_id, median_tpm, rsid, molecular_trait_id, chromosome_y, gene_start, gene_end, phenotype_id]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#this is how it would look!\n",
    "#now here you filter the values to only look at the peaks that have the variant in them\n",
    "merged_dataframe_n = merged_dataframe[((merged_dataframe[\"position\"].astype(int) >= merged_dataframe[\"gene_start\"].astype(int)) & (merged_dataframe[\"position\"].astype(int) <= merged_dataframe[\"gene_end\"].astype(int)))]\n",
    "merged_dataframe_n #in this case just the first 100 rows don't have anything!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# doing it for all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#idea is we'll save one that has only the peaks with the variant in them, then a separate one that has all of them\n",
    "data_sus = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(234077507, 20)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_sus.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame in RAM: 33931.52 MB\n"
     ]
    }
   ],
   "source": [
    "mem = data_sus.memory_usage().sum()  # in bytes\n",
    "print(f\"DataFrame in RAM: {mem/1024**2:.2f} MB\") #yeah about right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_72900/3940714587.py:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  peaks_data = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018_peak_metadata.tsv', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249196</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56861329</td>\n",
       "      <td>56862459</td>\n",
       "      <td>Interval_249197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56863235</td>\n",
       "      <td>56864760</td>\n",
       "      <td>Interval_249198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249198</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56865516</td>\n",
       "      <td>56874990</td>\n",
       "      <td>Interval_249199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249199</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56876239</td>\n",
       "      <td>56880531</td>\n",
       "      <td>Interval_249200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249200</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56884100</td>\n",
       "      <td>56885316</td>\n",
       "      <td>Interval_249201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249201 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome  gene_start  gene_end     phenotype_id\n",
       "0            chr1       10034     10617       Interval_1\n",
       "1            chr1       28924     29435       Interval_2\n",
       "2            chr1      180749    181528       Interval_3\n",
       "3            chr1      182645    184537       Interval_4\n",
       "4            chr1      186367    187110       Interval_5\n",
       "...           ...         ...       ...              ...\n",
       "249196       chrY    56861329  56862459  Interval_249197\n",
       "249197       chrY    56863235  56864760  Interval_249198\n",
       "249198       chrY    56865516  56874990  Interval_249199\n",
       "249199       chrY    56876239  56880531  Interval_249200\n",
       "249200       chrY    56884100  56885316  Interval_249201\n",
       "\n",
       "[249201 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now let's do the other operations\n",
    "peaks_data = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018_peak_metadata.tsv', sep='\\t')\n",
    "peaks_data = peaks_data[[\"chromosome\", \"gene_start\", \"gene_end\", \"phenotype_pos\", \"phenotype_id\"]]\n",
    "peaks_data[\"chromosome\"] = \"chr\"+peaks_data[\"chromosome\"] .astype(str)\n",
    "peaks_data = peaks_data[['chromosome', 'gene_start', 'gene_end', \"phenotype_id\"]]\n",
    "peaks_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_sus[\"position0\"] = data_sus[\"position\"] - 1\n",
    "data_sus = data_sus[['chromosome', 'position0', 'position', 'ref', 'alt', 'variant',\n",
    "       'ma_samples', 'maf', 'pvalue', 'beta', 'se', 'type', 'ac', 'an', 'r2',\n",
    "       'molecular_trait_object_id', 'gene_id', 'median_tpm', 'rsid',\n",
    "       'molecular_trait_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_dataframe = data_sus.merge(peaks_data, right_on=['phenotype_id'], left_on=[\"molecular_trait_id\"], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(234077507, 24)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_dataframe.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_x</th>\n",
       "      <th>position0</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>...</th>\n",
       "      <th>r2</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.071285</td>\n",
       "      <td>-0.328458</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.287697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.480256</td>\n",
       "      <td>0.108094</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.805997</td>\n",
       "      <td>-0.042729</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.206241</td>\n",
       "      <td>0.222002</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.970258</td>\n",
       "      <td>0.005194</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>190802</td>\n",
       "      <td>191922</td>\n",
       "      <td>Interval_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.538671</td>\n",
       "      <td>0.163632</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.251985</td>\n",
       "      <td>-0.346142</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.074580</td>\n",
       "      <td>0.391881</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>14932</td>\n",
       "      <td>14933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14933_G_A</td>\n",
       "      <td>13</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.023430</td>\n",
       "      <td>0.561551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome_x  position0  position ref alt         variant  ma_samples  \\\n",
       "0             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "1             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "2             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "3             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "4             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "5             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "6             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "7             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "8             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "9             1      14932     14933   G   A  chr1_14933_G_A          13   \n",
       "\n",
       "        maf    pvalue      beta  ...  r2 molecular_trait_object_id  \\\n",
       "0  0.263736  0.071285 -0.328458  ... NaN                Interval_1   \n",
       "1  0.263736  0.167217  0.287697  ... NaN                Interval_2   \n",
       "2  0.263736  0.480256  0.108094  ... NaN                Interval_3   \n",
       "3  0.263736  0.805997 -0.042729  ... NaN                Interval_4   \n",
       "4  0.263736  0.206241  0.222002  ... NaN                Interval_5   \n",
       "5  0.263736  0.970258  0.005194  ... NaN                Interval_6   \n",
       "6  0.071429  0.538671  0.163632  ... NaN                Interval_1   \n",
       "7  0.071429  0.251985 -0.346142  ... NaN                Interval_2   \n",
       "8  0.071429  0.074580  0.391881  ... NaN                Interval_3   \n",
       "9  0.071429  0.023430  0.561551  ... NaN                Interval_4   \n",
       "\n",
       "      gene_id  median_tpm         rsid molecular_trait_id chromosome_y  \\\n",
       "0  Interval_1         NaN  rs546169444         Interval_1         chr1   \n",
       "1  Interval_2         NaN  rs546169444         Interval_2         chr1   \n",
       "2  Interval_3         NaN  rs546169444         Interval_3         chr1   \n",
       "3  Interval_4         NaN  rs546169444         Interval_4         chr1   \n",
       "4  Interval_5         NaN  rs546169444         Interval_5         chr1   \n",
       "5  Interval_6         NaN  rs546169444         Interval_6         chr1   \n",
       "6  Interval_1         NaN  rs199856693         Interval_1         chr1   \n",
       "7  Interval_2         NaN  rs199856693         Interval_2         chr1   \n",
       "8  Interval_3         NaN  rs199856693         Interval_3         chr1   \n",
       "9  Interval_4         NaN  rs199856693         Interval_4         chr1   \n",
       "\n",
       "   gene_start gene_end phenotype_id  \n",
       "0       10034    10617   Interval_1  \n",
       "1       28924    29435   Interval_2  \n",
       "2      180749   181528   Interval_3  \n",
       "3      182645   184537   Interval_4  \n",
       "4      186367   187110   Interval_5  \n",
       "5      190802   191922   Interval_6  \n",
       "6       10034    10617   Interval_1  \n",
       "7       28924    29435   Interval_2  \n",
       "8      180749   181528   Interval_3  \n",
       "9      182645   184537   Interval_4  \n",
       "\n",
       "[10 rows x 24 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_dataframe.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_dataframe_n = merged_dataframe[((merged_dataframe[\"position\"].astype(int) >= merged_dataframe[\"gene_start\"].astype(int)) & (merged_dataframe[\"position\"].astype(int) <= merged_dataframe[\"gene_end\"].astype(int)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(806749, 24)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_dataframe_n.shape #slighly smaller since we removed variants that don't agree with bim files!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome_x', 'position0', 'position', 'ref', 'alt', 'variant',\n",
       "       'ma_samples', 'maf', 'pvalue', 'beta', 'se', 'type', 'ac', 'an', 'r2',\n",
       "       'molecular_trait_object_id', 'gene_id', 'median_tpm', 'rsid',\n",
       "       'molecular_trait_id', 'chromosome_y', 'gene_start', 'gene_end',\n",
       "       'phenotype_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_dataframe_n.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_dataframe_n.to_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset_peaks.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#and let's save out the one with all the peaks, but first find ones whose peak is within 524288//2 base pairs\n",
    "half_win = 524288 // 2\n",
    "mask = (\n",
    "    ((merged_dataframe['gene_end']   < merged_dataframe['position']) &\n",
    "     (merged_dataframe['gene_start'] >= merged_dataframe['position'] - half_win)\n",
    "    )\n",
    "    |\n",
    "    ((merged_dataframe['gene_start'] > merged_dataframe['position']) &\n",
    "     (merged_dataframe['gene_end']   <= merged_dataframe['position'] + half_win)\n",
    "    )\n",
    ")\n",
    "\n",
    "subset_peaks = merged_dataframe[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(233270758, 24)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_peaks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_x</th>\n",
       "      <th>position0</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>...</th>\n",
       "      <th>r2</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.071285</td>\n",
       "      <td>-0.328458</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.287697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.480256</td>\n",
       "      <td>0.108094</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.805997</td>\n",
       "      <td>-0.042729</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14463</td>\n",
       "      <td>14464</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.206241</td>\n",
       "      <td>0.222002</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome_x  position0  position ref alt         variant  ma_samples  \\\n",
       "0             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "1             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "2             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "3             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "4             1      14463     14464   A   T  chr1_14464_A_T          48   \n",
       "\n",
       "        maf    pvalue      beta  ...  r2 molecular_trait_object_id  \\\n",
       "0  0.263736  0.071285 -0.328458  ... NaN                Interval_1   \n",
       "1  0.263736  0.167217  0.287697  ... NaN                Interval_2   \n",
       "2  0.263736  0.480256  0.108094  ... NaN                Interval_3   \n",
       "3  0.263736  0.805997 -0.042729  ... NaN                Interval_4   \n",
       "4  0.263736  0.206241  0.222002  ... NaN                Interval_5   \n",
       "\n",
       "      gene_id  median_tpm         rsid molecular_trait_id chromosome_y  \\\n",
       "0  Interval_1         NaN  rs546169444         Interval_1         chr1   \n",
       "1  Interval_2         NaN  rs546169444         Interval_2         chr1   \n",
       "2  Interval_3         NaN  rs546169444         Interval_3         chr1   \n",
       "3  Interval_4         NaN  rs546169444         Interval_4         chr1   \n",
       "4  Interval_5         NaN  rs546169444         Interval_5         chr1   \n",
       "\n",
       "   gene_start gene_end phenotype_id  \n",
       "0       10034    10617   Interval_1  \n",
       "1       28924    29435   Interval_2  \n",
       "2      180749   181528   Interval_3  \n",
       "3      182645   184537   Interval_4  \n",
       "4      186367   187110   Interval_5  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_peaks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#let's see if we want to remove the position or position 0 column, probably position\n",
    "import numpy as np\n",
    "genome_np = np.load('/data1/lesliec/sarthak/data/chrombpnet_test/hg38_tokenized.npz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([7, 9], dtype=int8)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_np['chr1'][14463:14463+2] #yeah we definitely want position 0...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chromosome_x                              1\n",
       "position0                             16948\n",
       "position                              16949\n",
       "ref                                       A\n",
       "alt                                       C\n",
       "variant                      chr1_16949_A_C\n",
       "ma_samples                               67\n",
       "maf                                0.368132\n",
       "pvalue                             0.556453\n",
       "beta                               0.143291\n",
       "se                                 0.242579\n",
       "type                                    SNP\n",
       "ac                                       67\n",
       "an                                      182\n",
       "r2                                      NaN\n",
       "molecular_trait_object_id        Interval_2\n",
       "gene_id                          Interval_2\n",
       "median_tpm                              NaN\n",
       "rsid                            rs199745162\n",
       "molecular_trait_id               Interval_2\n",
       "chromosome_y                           chr1\n",
       "gene_start                            28924\n",
       "gene_end                              29435\n",
       "phenotype_id                     Interval_2\n",
       "Name: 13, dtype: object"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#let's check one more example\n",
    "subset_peaks.iloc[13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_np['chr1'][16948]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#let's drop the position ahd chromosome_x columns\n",
    "subset_peaks = subset_peaks.drop(columns=[\"position\", \"chromosome_x\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(233270758, 22)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_peaks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[38], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43msubset_peaks\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset_peaks_nearby.tsv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/data1/lesliec/sarthak/caduceus/.pixi/envs/default/lib/python3.11/site-packages/pandas/core/generic.py:3902\u001b[0m, in \u001b[0;36mNDFrame.to_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[0m\n\u001b[1;32m   3891\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ABCDataFrame) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mto_frame()\n\u001b[1;32m   3893\u001b[0m formatter \u001b[38;5;241m=\u001b[39m DataFrameFormatter(\n\u001b[1;32m   3894\u001b[0m     frame\u001b[38;5;241m=\u001b[39mdf,\n\u001b[1;32m   3895\u001b[0m     header\u001b[38;5;241m=\u001b[39mheader,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3899\u001b[0m     decimal\u001b[38;5;241m=\u001b[39mdecimal,\n\u001b[1;32m   3900\u001b[0m )\n\u001b[0;32m-> 3902\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameRenderer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformatter\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3903\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3904\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlineterminator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3905\u001b[0m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3906\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3907\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3908\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3909\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquoting\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquoting\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3910\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3911\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex_label\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_label\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3912\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3913\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3914\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquotechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquotechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3915\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3916\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdoublequote\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3917\u001b[0m \u001b[43m    \u001b[49m\u001b[43mescapechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mescapechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3918\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3919\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/data1/lesliec/sarthak/caduceus/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/formats/format.py:1152\u001b[0m, in \u001b[0;36mDataFrameRenderer.to_csv\u001b[0;34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[0m\n\u001b[1;32m   1131\u001b[0m     created_buffer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m   1133\u001b[0m csv_formatter \u001b[38;5;241m=\u001b[39m CSVFormatter(\n\u001b[1;32m   1134\u001b[0m     path_or_buf\u001b[38;5;241m=\u001b[39mpath_or_buf,\n\u001b[1;32m   1135\u001b[0m     lineterminator\u001b[38;5;241m=\u001b[39mlineterminator,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1150\u001b[0m     formatter\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfmt,\n\u001b[1;32m   1151\u001b[0m )\n\u001b[0;32m-> 1152\u001b[0m \u001b[43mcsv_formatter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1154\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m created_buffer:\n\u001b[1;32m   1155\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, StringIO)\n",
      "File \u001b[0;32m/data1/lesliec/sarthak/caduceus/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/formats/csvs.py:266\u001b[0m, in \u001b[0;36mCSVFormatter.save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    247\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m get_handle(\n\u001b[1;32m    248\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilepath_or_buffer,\n\u001b[1;32m    249\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmode,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    254\u001b[0m ) \u001b[38;5;28;01mas\u001b[39;00m handles:\n\u001b[1;32m    255\u001b[0m     \u001b[38;5;66;03m# Note: self.encoding is irrelevant here\u001b[39;00m\n\u001b[1;32m    256\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwriter \u001b[38;5;241m=\u001b[39m csvlib\u001b[38;5;241m.\u001b[39mwriter(\n\u001b[1;32m    257\u001b[0m         handles\u001b[38;5;241m.\u001b[39mhandle,\n\u001b[1;32m    258\u001b[0m         lineterminator\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlineterminator,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    263\u001b[0m         quotechar\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mquotechar,\n\u001b[1;32m    264\u001b[0m     )\n\u001b[0;32m--> 266\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_save\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/data1/lesliec/sarthak/caduceus/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/formats/csvs.py:271\u001b[0m, in \u001b[0;36mCSVFormatter._save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    269\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_need_to_save_header:\n\u001b[1;32m    270\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_save_header()\n\u001b[0;32m--> 271\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_save_body\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/data1/lesliec/sarthak/caduceus/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/formats/csvs.py:309\u001b[0m, in \u001b[0;36mCSVFormatter._save_body\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    307\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m start_i \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m end_i:\n\u001b[1;32m    308\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[0;32m--> 309\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_save_chunk\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstart_i\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mend_i\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/data1/lesliec/sarthak/caduceus/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/formats/csvs.py:320\u001b[0m, in \u001b[0;36mCSVFormatter._save_chunk\u001b[0;34m(self, start_i, end_i)\u001b[0m\n\u001b[1;32m    317\u001b[0m data \u001b[38;5;241m=\u001b[39m [res\u001b[38;5;241m.\u001b[39miget_values(i) \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(res\u001b[38;5;241m.\u001b[39mitems))]\n\u001b[1;32m    319\u001b[0m ix \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdata_index[slicer]\u001b[38;5;241m.\u001b[39m_format_native_types(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_number_format)\n\u001b[0;32m--> 320\u001b[0m \u001b[43mlibwriters\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_csv_rows\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    321\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    322\u001b[0m \u001b[43m    \u001b[49m\u001b[43mix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    323\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnlevels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    324\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcols\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    325\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwriter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    326\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32mwriters.pyx:72\u001b[0m, in \u001b[0;36mpandas._libs.writers.write_csv_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "subset_peaks.to_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset_peaks_nearby.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position0</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>type</th>\n",
       "      <th>...</th>\n",
       "      <th>r2</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.071285</td>\n",
       "      <td>-0.328458</td>\n",
       "      <td>0.179592</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.287697</td>\n",
       "      <td>0.206328</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.480256</td>\n",
       "      <td>0.108094</td>\n",
       "      <td>0.152387</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.805997</td>\n",
       "      <td>-0.042729</td>\n",
       "      <td>0.173386</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.206241</td>\n",
       "      <td>0.222002</td>\n",
       "      <td>0.174157</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   position0 ref alt         variant  ma_samples       maf    pvalue  \\\n",
       "0      14463   A   T  chr1_14464_A_T          48  0.263736  0.071285   \n",
       "1      14463   A   T  chr1_14464_A_T          48  0.263736  0.167217   \n",
       "2      14463   A   T  chr1_14464_A_T          48  0.263736  0.480256   \n",
       "3      14463   A   T  chr1_14464_A_T          48  0.263736  0.805997   \n",
       "4      14463   A   T  chr1_14464_A_T          48  0.263736  0.206241   \n",
       "\n",
       "       beta        se type  ...  r2  molecular_trait_object_id     gene_id  \\\n",
       "0 -0.328458  0.179592  SNP  ... NaN                 Interval_1  Interval_1   \n",
       "1  0.287697  0.206328  SNP  ... NaN                 Interval_2  Interval_2   \n",
       "2  0.108094  0.152387  SNP  ... NaN                 Interval_3  Interval_3   \n",
       "3 -0.042729  0.173386  SNP  ... NaN                 Interval_4  Interval_4   \n",
       "4  0.222002  0.174157  SNP  ... NaN                 Interval_5  Interval_5   \n",
       "\n",
       "  median_tpm         rsid  molecular_trait_id chromosome_y gene_start  \\\n",
       "0        NaN  rs546169444          Interval_1         chr1      10034   \n",
       "1        NaN  rs546169444          Interval_2         chr1      28924   \n",
       "2        NaN  rs546169444          Interval_3         chr1     180749   \n",
       "3        NaN  rs546169444          Interval_4         chr1     182645   \n",
       "4        NaN  rs546169444          Interval_5         chr1     186367   \n",
       "\n",
       "  gene_end  phenotype_id  \n",
       "0    10617    Interval_1  \n",
       "1    29435    Interval_2  \n",
       "2   181528    Interval_3  \n",
       "3   184537    Interval_4  \n",
       "4   187110    Interval_5  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the final thing we will do is subset it to one row per variant but then a list of peaks that are for that variant\n",
    "subset_peaks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan])"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#let's see if the r2 column has any useful information\n",
    "subset_peaks[\"r2\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCannot execute code, session has been disposed. Please try restarting the Kernel."
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCannot execute code, session has been disposed. Please try restarting the Kernel. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "#let's drop that too\n",
    "subset_peaks = subset_peaks.drop(columns=[\"r2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_peaks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#let's see if molecular trait object id column is identical to gene id\n",
    "identical = subset_peaks[\"molecular_trait_object_id\"].equals(subset_peaks[\"gene_id\"])\n",
    "identical"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# redoing the big one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_103818/1299764431.py:4: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  peaks_data = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018_peak_metadata.tsv', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "#we are redoing it because we have a lot of editing and processing to do!\n",
    "import pandas as pd\n",
    "data_sus = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset.tsv', sep='\\t')\n",
    "peaks_data = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018_peak_metadata.tsv', sep='\\t')\n",
    "peaks_data = peaks_data[[\"chromosome\", \"gene_start\", \"gene_end\", \"phenotype_pos\", \"phenotype_id\"]]\n",
    "peaks_data[\"chromosome\"] = \"chr\"+peaks_data[\"chromosome\"] .astype(str)\n",
    "peaks_data = peaks_data[['chromosome', 'gene_start', 'gene_end', \"phenotype_id\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_sus[\"position0\"] = data_sus[\"position\"] - 1\n",
    "data_sus = data_sus[['chromosome', 'position0', 'position', 'ref', 'alt', 'variant',\n",
    "       'ma_samples', 'maf', 'pvalue', 'beta', 'se', 'type', 'ac', 'an', 'r2',\n",
    "       'molecular_trait_object_id', 'gene_id', 'median_tpm', 'rsid',\n",
    "       'molecular_trait_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_dataframe = data_sus.merge(peaks_data, right_on=['phenotype_id'], left_on=[\"molecular_trait_id\"], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "half_win = 524288 // 2\n",
    "mask = (\n",
    "    ((merged_dataframe['gene_end']   < merged_dataframe['position']) &\n",
    "     (merged_dataframe['gene_start'] >= merged_dataframe['position'] - half_win)\n",
    "    )\n",
    "    |\n",
    "    ((merged_dataframe['gene_start'] > merged_dataframe['position']) &\n",
    "     (merged_dataframe['gene_end']   <= merged_dataframe['position'] + half_win)\n",
    "    )\n",
    ")\n",
    "\n",
    "subset_peaks = merged_dataframe[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SNP'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#let's see if we need to drop anymore peaks first\n",
    "#the type column should just be snp\n",
    "subset_peaks[\"type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the median tpm should be nan as well\n",
    "subset_peaks[\"median_tpm\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "#these column might all be identical, molecular_trait_object_id, gene_id, phenotype_id and molecular_trait_id\n",
    "identical = subset_peaks[\"gene_id\"].equals(subset_peaks[\"molecular_trait_object_id\"])\n",
    "print(identical)\n",
    "identical = subset_peaks[\"gene_id\"].equals(subset_peaks[\"phenotype_id\"])\n",
    "print(identical)\n",
    "identical = subset_peaks[\"gene_id\"].equals(subset_peaks[\"molecular_trait_id\"])\n",
    "print(identical)\n",
    "#they are all identical, so we can drop all of them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(233270758, 16)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_peaks = subset_peaks.drop(columns=[\"position\", \"chromosome_x\", \"r2\", \"type\", \"median_tpm\", \"phenotype_id\", \"molecular_trait_object_id\", \"molecular_trait_id\"])\n",
    "subset_peaks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position0</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>ac</th>\n",
       "      <th>an</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.071285</td>\n",
       "      <td>-0.328458</td>\n",
       "      <td>0.179592</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>Interval_1</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.287697</td>\n",
       "      <td>0.206328</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>Interval_2</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.480256</td>\n",
       "      <td>0.108094</td>\n",
       "      <td>0.152387</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>Interval_3</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.805997</td>\n",
       "      <td>-0.042729</td>\n",
       "      <td>0.173386</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>Interval_4</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14463</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14464_A_T</td>\n",
       "      <td>48</td>\n",
       "      <td>0.263736</td>\n",
       "      <td>0.206241</td>\n",
       "      <td>0.222002</td>\n",
       "      <td>0.174157</td>\n",
       "      <td>48</td>\n",
       "      <td>182</td>\n",
       "      <td>Interval_5</td>\n",
       "      <td>rs546169444</td>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position0 ref alt         variant  ma_samples       maf    pvalue  \\\n",
       "0      14463   A   T  chr1_14464_A_T          48  0.263736  0.071285   \n",
       "1      14463   A   T  chr1_14464_A_T          48  0.263736  0.167217   \n",
       "2      14463   A   T  chr1_14464_A_T          48  0.263736  0.480256   \n",
       "3      14463   A   T  chr1_14464_A_T          48  0.263736  0.805997   \n",
       "4      14463   A   T  chr1_14464_A_T          48  0.263736  0.206241   \n",
       "\n",
       "       beta        se  ac   an     gene_id         rsid chromosome_y  \\\n",
       "0 -0.328458  0.179592  48  182  Interval_1  rs546169444         chr1   \n",
       "1  0.287697  0.206328  48  182  Interval_2  rs546169444         chr1   \n",
       "2  0.108094  0.152387  48  182  Interval_3  rs546169444         chr1   \n",
       "3 -0.042729  0.173386  48  182  Interval_4  rs546169444         chr1   \n",
       "4  0.222002  0.174157  48  182  Interval_5  rs546169444         chr1   \n",
       "\n",
       "   gene_start  gene_end  \n",
       "0       10034     10617  \n",
       "1       28924     29435  \n",
       "2      180749    181528  \n",
       "3      182645    184537  \n",
       "4      186367    187110  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_peaks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>phenotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10034</td>\n",
       "      <td>10617</td>\n",
       "      <td>Interval_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28924</td>\n",
       "      <td>29435</td>\n",
       "      <td>Interval_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>180749</td>\n",
       "      <td>181528</td>\n",
       "      <td>Interval_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182645</td>\n",
       "      <td>184537</td>\n",
       "      <td>Interval_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>186367</td>\n",
       "      <td>187110</td>\n",
       "      <td>Interval_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249196</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56861329</td>\n",
       "      <td>56862459</td>\n",
       "      <td>Interval_249197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56863235</td>\n",
       "      <td>56864760</td>\n",
       "      <td>Interval_249198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249198</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56865516</td>\n",
       "      <td>56874990</td>\n",
       "      <td>Interval_249199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249199</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56876239</td>\n",
       "      <td>56880531</td>\n",
       "      <td>Interval_249200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249200</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56884100</td>\n",
       "      <td>56885316</td>\n",
       "      <td>Interval_249201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249201 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome  gene_start  gene_end     phenotype_id\n",
       "0            chr1       10034     10617       Interval_1\n",
       "1            chr1       28924     29435       Interval_2\n",
       "2            chr1      180749    181528       Interval_3\n",
       "3            chr1      182645    184537       Interval_4\n",
       "4            chr1      186367    187110       Interval_5\n",
       "...           ...         ...       ...              ...\n",
       "249196       chrY    56861329  56862459  Interval_249197\n",
       "249197       chrY    56863235  56864760  Interval_249198\n",
       "249198       chrY    56865516  56874990  Interval_249199\n",
       "249199       chrY    56876239  56880531  Interval_249200\n",
       "249200       chrY    56884100  56885316  Interval_249201\n",
       "\n",
       "[249201 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#the other thing we'll do is remove ma_samples, gene start and gene end columns\n",
    "subset_peaks = subset_peaks.drop(columns=[\"ma_samples\", \"gene_start\", \"gene_end\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6508629, 12)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now let's group by and then make lists\n",
    "#can find number of elements in the group by checking length of the list!\n",
    "\n",
    "agg_dict = {\n",
    "    'position0':     'first',\n",
    "    'ref':           'first',\n",
    "    'alt':           'first',\n",
    "    'maf':           'first',\n",
    "    'ac':            'first',\n",
    "    'an':            'first',\n",
    "    'chromosome_y':  'first',\n",
    "    'rsid':          'first',\n",
    "    'pvalue':        list,\n",
    "    'beta':          list,\n",
    "    'gene_id':       list,\n",
    "}\n",
    "\n",
    "grouped = subset_peaks \\\n",
    "    .groupby('variant', as_index=False) \\\n",
    "    .agg(agg_dict)\n",
    "grouped.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>position0</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>maf</th>\n",
       "      <th>ac</th>\n",
       "      <th>an</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>rsid</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100000235_C_T</td>\n",
       "      <td>100000234</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.340659</td>\n",
       "      <td>62</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11596870</td>\n",
       "      <td>[0.719412, 0.113877, 0.953659, 0.00137416, 3.8...</td>\n",
       "      <td>[0.0512266, -0.237318, 0.00746471, -0.353161, ...</td>\n",
       "      <td>[Interval_26170, Interval_26171, Interval_2617...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_100000943_G_A</td>\n",
       "      <td>100000942</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>14</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190359</td>\n",
       "      <td>[0.0224591, 0.0439213, 0.129788, 0.584645, 0.5...</td>\n",
       "      <td>[-0.622646, 0.585245, -0.375654, -0.12116, -0....</td>\n",
       "      <td>[Interval_26170, Interval_26171, Interval_2617...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_100000979_T_C</td>\n",
       "      <td>100000978</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.032967</td>\n",
       "      <td>6</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190360</td>\n",
       "      <td>[0.0172942, 0.735536, 0.029347, 0.0188991, 0.4...</td>\n",
       "      <td>[-0.95319, 0.146098, -0.787944, -0.752098, 0.3...</td>\n",
       "      <td>[Interval_26170, Interval_26171, Interval_2617...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_100002012_T_C</td>\n",
       "      <td>100002011</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.043956</td>\n",
       "      <td>8</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190362</td>\n",
       "      <td>[0.412635, 0.0255791, 0.955902, 0.199178, 0.15...</td>\n",
       "      <td>[-0.283286, 0.807535, -0.0172548, 0.353885, -0...</td>\n",
       "      <td>[Interval_26170, Interval_26171, Interval_2617...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10_100002628_A_C</td>\n",
       "      <td>100002627</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>84</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190363</td>\n",
       "      <td>[0.143933, 0.566345, 0.329146, 0.000515478, 0....</td>\n",
       "      <td>[-0.198788, -0.0832435, -0.119994, -0.36588, 0...</td>\n",
       "      <td>[Interval_26170, Interval_26171, Interval_2617...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               variant  position0 ref alt       maf  ac   an chromosome_y  \\\n",
       "0  chr10_100000235_C_T  100000234   C   T  0.340659  62  182        chr10   \n",
       "1  chr10_100000943_G_A  100000942   G   A  0.076923  14  182        chr10   \n",
       "2  chr10_100000979_T_C  100000978   T   C  0.032967   6  182        chr10   \n",
       "3  chr10_100002012_T_C  100002011   T   C  0.043956   8  182        chr10   \n",
       "4  chr10_100002628_A_C  100002627   A   C  0.461538  84  182        chr10   \n",
       "\n",
       "         rsid                                             pvalue  \\\n",
       "0  rs11596870  [0.719412, 0.113877, 0.953659, 0.00137416, 3.8...   \n",
       "1  rs11190359  [0.0224591, 0.0439213, 0.129788, 0.584645, 0.5...   \n",
       "2  rs11190360  [0.0172942, 0.735536, 0.029347, 0.0188991, 0.4...   \n",
       "3  rs11190362  [0.412635, 0.0255791, 0.955902, 0.199178, 0.15...   \n",
       "4  rs11190363  [0.143933, 0.566345, 0.329146, 0.000515478, 0....   \n",
       "\n",
       "                                                beta  \\\n",
       "0  [0.0512266, -0.237318, 0.00746471, -0.353161, ...   \n",
       "1  [-0.622646, 0.585245, -0.375654, -0.12116, -0....   \n",
       "2  [-0.95319, 0.146098, -0.787944, -0.752098, 0.3...   \n",
       "3  [-0.283286, 0.807535, -0.0172548, 0.353885, -0...   \n",
       "4  [-0.198788, -0.0832435, -0.119994, -0.36588, 0...   \n",
       "\n",
       "                                             gene_id  \n",
       "0  [Interval_26170, Interval_26171, Interval_2617...  \n",
       "1  [Interval_26170, Interval_26171, Interval_2617...  \n",
       "2  [Interval_26170, Interval_26171, Interval_2617...  \n",
       "3  [Interval_26170, Interval_26171, Interval_2617...  \n",
       "4  [Interval_26170, Interval_26171, Interval_2617...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now let's save this out!\n",
    "grouped.to_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset_peaks_nearby.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>position0</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>maf</th>\n",
       "      <th>ac</th>\n",
       "      <th>an</th>\n",
       "      <th>chromosome_y</th>\n",
       "      <th>rsid</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100000235_C_T</td>\n",
       "      <td>100000234</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.340659</td>\n",
       "      <td>62</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11596870</td>\n",
       "      <td>[0.719412, 0.113877, 0.953659, 0.00137416, 3.8...</td>\n",
       "      <td>[0.0512266, -0.237318, 0.00746471, -0.353161, ...</td>\n",
       "      <td>['Interval_26170', 'Interval_26171', 'Interval...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_100000943_G_A</td>\n",
       "      <td>100000942</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>14</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190359</td>\n",
       "      <td>[0.0224591, 0.0439213, 0.129788, 0.584645, 0.5...</td>\n",
       "      <td>[-0.622646, 0.585245, -0.375654, -0.12116, -0....</td>\n",
       "      <td>['Interval_26170', 'Interval_26171', 'Interval...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_100000979_T_C</td>\n",
       "      <td>100000978</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.032967</td>\n",
       "      <td>6</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190360</td>\n",
       "      <td>[0.0172942, 0.735536, 0.029347, 0.0188991, 0.4...</td>\n",
       "      <td>[-0.95319, 0.146098, -0.787944, -0.752098, 0.3...</td>\n",
       "      <td>['Interval_26170', 'Interval_26171', 'Interval...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_100002012_T_C</td>\n",
       "      <td>100002011</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.043956</td>\n",
       "      <td>8</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190362</td>\n",
       "      <td>[0.412635, 0.0255791, 0.955902, 0.199178, 0.15...</td>\n",
       "      <td>[-0.283286, 0.807535, -0.0172548, 0.353885, -0...</td>\n",
       "      <td>['Interval_26170', 'Interval_26171', 'Interval...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10_100002628_A_C</td>\n",
       "      <td>100002627</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>84</td>\n",
       "      <td>182</td>\n",
       "      <td>chr10</td>\n",
       "      <td>rs11190363</td>\n",
       "      <td>[0.143933, 0.566345, 0.329146, 0.000515478, 0....</td>\n",
       "      <td>[-0.198788, -0.0832435, -0.119994, -0.36588, 0...</td>\n",
       "      <td>['Interval_26170', 'Interval_26171', 'Interval...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               variant  position0 ref alt       maf  ac   an chromosome_y  \\\n",
       "0  chr10_100000235_C_T  100000234   C   T  0.340659  62  182        chr10   \n",
       "1  chr10_100000943_G_A  100000942   G   A  0.076923  14  182        chr10   \n",
       "2  chr10_100000979_T_C  100000978   T   C  0.032967   6  182        chr10   \n",
       "3  chr10_100002012_T_C  100002011   T   C  0.043956   8  182        chr10   \n",
       "4  chr10_100002628_A_C  100002627   A   C  0.461538  84  182        chr10   \n",
       "\n",
       "         rsid                                             pvalue  \\\n",
       "0  rs11596870  [0.719412, 0.113877, 0.953659, 0.00137416, 3.8...   \n",
       "1  rs11190359  [0.0224591, 0.0439213, 0.129788, 0.584645, 0.5...   \n",
       "2  rs11190360  [0.0172942, 0.735536, 0.029347, 0.0188991, 0.4...   \n",
       "3  rs11190362  [0.412635, 0.0255791, 0.955902, 0.199178, 0.15...   \n",
       "4  rs11190363  [0.143933, 0.566345, 0.329146, 0.000515478, 0....   \n",
       "\n",
       "                                                beta  \\\n",
       "0  [0.0512266, -0.237318, 0.00746471, -0.353161, ...   \n",
       "1  [-0.622646, 0.585245, -0.375654, -0.12116, -0....   \n",
       "2  [-0.95319, 0.146098, -0.787944, -0.752098, 0.3...   \n",
       "3  [-0.283286, 0.807535, -0.0172548, 0.353885, -0...   \n",
       "4  [-0.198788, -0.0832435, -0.119994, -0.36588, 0...   \n",
       "\n",
       "                                             gene_id  \n",
       "0  ['Interval_26170', 'Interval_26171', 'Interval...  \n",
       "1  ['Interval_26170', 'Interval_26171', 'Interval...  \n",
       "2  ['Interval_26170', 'Interval_26171', 'Interval...  \n",
       "3  ['Interval_26170', 'Interval_26171', 'Interval...  \n",
       "4  ['Interval_26170', 'Interval_26171', 'Interval...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#let's load it in and make sure it's good\n",
    "grouped2 = pd.read_csv('/data1/lesliec/sarthak/data/joint_playground/caQTL/QTD100018.all_subset_peaks_nearby.tsv', sep='\\t', nrows=100)\n",
    "grouped2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#yeah it looks fine!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
